File: 2416-sum-of-prefix-scores-of-strings/2416-sum-of-prefix-scores-of-strings.js
/**
 * @param {string[]} words
 * @return {number[]}
 */
var sumPrefixScores = function(words) {
  const trie = new Trie();
  
  words.forEach((word) => trie.insert(word));
  
  return words.map((word) => trie.getScore(word));
};

class Trie {
  constructor() {
    this.root = new TrieNode();
  }
  
  insert(word) {
    let node = this.root;
    
    for (let ch of word) {
      if (!node.next[ch]) node.next[ch] = new TrieNode();
      
      node = node.next[ch];
      node.count++;
    }
  }
  
  getScore(word) {
    let node = this.root;
    let score = 0;
    
    for (let ch of word) {
      node = node.next[ch];
      score += node.count;
    }
    
    return score;
  }
}

class TrieNode {
  constructor() {
    this.next = {};
    this.count = 0;
  }
}

File: 2416-sum-of-prefix-scores-of-strings/2416-sum-of-prefix-scores-of-strings.py
class TrieNode:
    def __init__(self):
        self.next = [None] * 26
        self.count = 0
        
  
class Trie:
    def __init__(self):
        self.root = TrieNode()
        
    def insert(self, word):
        node = self.root
        for ch in word:
            i = ord(ch) - ord('a')
            if not node.next[i]:
                node.next[i] = TrieNode()
            node = node.next[i]
            node.count += 1
    
    def getScore(self, word):
        node, score = self.root, 0
        for ch in word:
            i = ord(ch) - ord('a')
            node = node.next[i]
            score += node.count
        return score
      

class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        trie = Trie()

        for word in words:
            trie.insert(word)

        return [trie.getScore(word) for word in words]